core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,268 @@
+# coding: utf-8
+#
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for tracking the progress of the learner."""
+
+from core.domain import subscription_services
+from core.platform import models
+import utils
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+
+def add_exp_id_to_completed_list(user_id, exp_id):
+    """Adds the exploration id to the completed list of the user unless the
+    exploration has already been completed or has been created by the user.
+    It is also removed from the incomplete list (if present).
+
+    Args:
+        user_id: str. The id of the user who has completed the exploration.
+        exp_id: str. The id of the completed exploration.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    exploration_created_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    if (exp_id not in exploration_created_ids and
+            exp_id not in activities_completed_model.completed_exploration_ids):
```

Comment:
Ah, I see what you mean re user_services. The function there is actually not correct. An ExplorationUserData domain object ought to have been created.

For collection_services, that is correct. Remember I mentioned that get, create and save are the only things we should be using the storage layer objects for? (I forgot "delete", too, btw.) _save_collection is the function that converts a collection domain object to a collection model storage layer object, and is the third of these exceptions.

In fact, collection_services is a good model to follow in general, and so is exp_services. You'll notice that most of the functions operate with Collection and Exploration domain objects respectively and not the models from the datastore.

Does that help clarify?